File: client-ios/SerenadaCore/Sources/Call/IceCandidateSanitizer.swift
import Foundation

/// Drops blank-SDP candidates and normalizes blank `sdpMid` to nil so native
/// WebRTC falls back to `sdpMLineIndex`. Synthesizing a numeric mid from the
/// m-line index would mismatch remote SDPs that use named mids (e.g. `audio`),
/// so we preserve a missing mid rather than fabricate one.
internal func sanitizeIceCandidate(
    _ candidate: IceCandidatePayload,
    remoteCid: String,
    logger: SerenadaLogger? = nil
) -> IceCandidatePayload? {
    let trimmedCandidate = candidate.candidate.trimmingCharacters(in: .whitespacesAndNewlines)
    if trimmedCandidate.isEmpty {
        logger?.log(.warning, tag: "PeerConnection", "[\(remoteCid)] Dropping blank ICE candidate")
        return nil
    }

    let trimmedMid = candidate.sdpMid?.trimmingCharacters(in: .whitespacesAndNewlines)
    if let mid = trimmedMid, !mid.isEmpty {
        return candidate
    }
    return IceCandidatePayload(
        sdpMid: nil,
        sdpMLineIndex: candidate.sdpMLineIndex,
        candidate: candidate.candidate
    )
}

File: client-android/serenada-core/src/main/java/app/serenada/core/call/IceCandidateSanitizer.kt
package app.serenada.core.call

import app.serenada.core.SerenadaLogLevel
import app.serenada.core.SerenadaLogger
import org.webrtc.IceCandidate

/**
 * Drops blank SDP candidates and synthesizes `sdpMid` from `sdpMLineIndex` when
 * the remote omitted it. Native WebRTC rejects candidates with blank or
 * mismatched mid metadata, so we filter at the boundary before they reach the
 * peer connection (or the pending buffer).
 */
internal fun sanitizeIceCandidate(
    candidate: IceCandidate,
    remoteCid: String,
    logger: SerenadaLogger? = null,
): IceCandidate? {
    val candidateSdp = candidate.sdp?.takeIf { it.isNotBlank() }
    if (candidateSdp == null) {
        logger?.log(SerenadaLogLevel.WARNING, "PeerConnection", "[$remoteCid] Dropping blank ICE candidate")
        return null
    }
    val sdpMid = candidate.sdpMid?.takeIf { it.isNotBlank() }
    if (sdpMid != null) return candidate
    return IceCandidate(candidate.sdpMLineIndex.toString(), candidate.sdpMLineIndex, candidateSdp)
}
